packages/labs/context/src/lib/controllers/context-provider.ts

```diff
@@ -83,19 +83,50 @@ export class ContextProvider<T extends Context<unknown, unknown>>
     // The check on composedPath (as opposed to ev.target) is to cover cases
     // where the consumer is in the shadowDom of the provider (in which case,
     // event.target === this.host because of event retargeting).
-    if (ev.context !== this.context || ev.composedPath()[0] === this.host) {
+    const consumerHost = ev.composedPath()[0] as Element;
+    if (ev.context !== this.context || consumerHost === this.host) {
       return;
     }
     ev.stopPropagation();
-    this.addCallback(ev.callback, ev.subscribe);
+    this.addCallback(ev.callback, ev.subscribe, consumerHost);
+  };
+
+  public onProviderRequest = (
+    ev: ContextProviderEvent<Context<unknown, unknown>>
+  ): void => {
+    // Ignore events when the context doesn't match.
+    // Also, in case an element is a consumer AND a provider
+    // of the same context it shouldn't provide to itself.
+    // We use composedPath (as opposed to ev.target) to cover cases
+    // where the consumer is in the shadowDom of the provider (in which case,
+    // event.target === this.host because of event retargeting).
+    const childProviderHost = ev.composedPath()[0] as Element;
+    if (ev.context !== this.context || childProviderHost === this.host) {
+      return;
+    }
+    this.reparentSubscriptions(childProviderHost);
+    ev.stopPropagation();
   };
 
   private attachListeners() {
     this.host.addEventListener('context-request', this.onContextRequest);
+    this.host.addEventListener('context-provider', this.onProviderRequest);
   }
 
   hostConnected(): void {
     // emit an event to signal a provider is available for this context
-    this.host.dispatchEvent(new ContextProviderEvent(this.context));
+    const providerEvent = new ContextProviderEvent(this.context);
+    this.host.dispatchEvent(providerEvent);
+  }
+
+  private reparentSubscriptions(_childProviderHost: Element) {
+    for (const [callback, {consumerHost}] of this.callbacks) {
+      if (consumerHost === undefined) {
+        continue;
+      }
+      consumerHost.dispatchEvent(
```

Comment:
Nit: add a comment for the why of this logic. Dispatching this event will ensure that the context will get the nearest provider.

Reply:
inlined into onProviderRequest and added a comment